File: Algorithms_python/max_subarray_tree.py
import sys
input = sys.stdin.readline
INF = 10 ** 18

class Node:
    def __init__(self):
        self.s = 0
        self.m = -INF
        self.l = -INF
        self.r = -INF

    def __add__(self, other):
        ret = Node()
        ret.s = self.s + other.s
        ret.l = max(self.l, self.s + other.l)
        ret.r = max(other.r, self.r + other.s)
        ret.m = max(self.r + other.l, max(self.m, other.m))
        return ret

class SegmentTree():
    def __init__(self, n):
        self.n = 2 ** (n - 1).bit_length()
        self.data = [Node() for _ in range(2 * self.n)]

    def update(self, k, v):
        k += self.n - 1
        self.data[k].s = self.data[k].m = self.data[k].l = self.data[k].r = v
        while k > 0:
            k = (k - 1) // 2
            l = k * 2 + 1
            r = k * 2 + 2
            self.data[k].s = self.data[l].s + self.data[r].s
            self.data[k].l = max(self.data[l].l, self.data[l].s + self.data[r].l)
            self.data[k].r = max(self.data[r].r, self.data[l].r + self.data[r].s)
            self.data[k].m = max(self.data[l].r + self.data[r].l, max(self.data[l].m, self.data[r].m))

    def query(self, l, r):
        L = l + self.n
        R = r + self.n
        ret_l = Node()
        ret_r = Node()
        while L < R:
            if R & 1:
                R -= 1
                ret_r = self.data[R - 1] + ret_r
            if L & 1:
                ret_l = ret_l + self.data[L - 1]
                L += 1
            L >>= 1
            R >>= 1
        return (ret_l + ret_r).m

if __name__ == "__main__":
    N = int(input())
    Seg = SegmentTree(N)
    arr = list(map(int, input().split()))
    for i in range(N):
        Seg.update(i, arr[i])

    Q = int(input())
    for _ in range(Q):
        l, r = map(int, input().split())
        print(Seg.query(l - 1, r))

File: Algorithms_cpp/max_subarray_tree.cpp
#include <bits/stdc++.h>
using namespace std;

const int INF = 1e9;

struct Node{
    int s, m, l, r;
    Node() : s(0), m(-INF), l(-INF), r(-INF) { }
    Node operator+(Node &right) {
        Node ret;
        ret.s = s + right.s;
        ret.l = max(l, s + right.l);
        ret.r = max(right.r, r + right.s);
        ret.m = max(r + right.l, max(m, right.m));
        return ret;
    }
};

struct SegmentTree{
    vector<Node> data;
    int n;
    SegmentTree(int size) {
        int p = 1;
        while (p < size) p *= 2;
        data.resize(p * 2);
        n = p;
    }

    void set(int i, int v) {
        i += n - 1;
        data[i].s = v;
        data[i].l = v;
        data[i].r = v;
        data[i].m = v;
    }

    void build() {
        for (int i = n - 2; i >= 0; i--) {
            data[i] = data[i * 2 + 1] + data[i * 2 + 2];
        }
    }

    void update(int i, int v) {
        i += n - 1;
        set(i, v);
        while (i > 0) {
            i = (i - 1) / 2;
            data[i] = data[i * 2 + 1] + data[i * 2 + 2];
        }
    }

    int query(int l, int r){
        l += n;
        r += n + 1;
        Node ret_l = Node();
        Node ret_r = Node();
        while (l < r) {
            if (r & 1) {
                r--;
                ret_r = data[r - 1] + ret_r;
            }
            if (l & 1) {
                ret_l = ret_l + data[l - 1];
                l++;
            }
            l >>= 1;
            r >>= 1;
        }
        return (ret_l + ret_r).m;
    }
};

int main() {
    cin.tie(nullptr);
    cout.tie(nullptr);
    ios::sync_with_stdio(false);
    int N;
    cin >> N;
    SegmentTree Seg = SegmentTree(N);
    for (int i = 0; i < N; i++) {
        int v;
        cin >> v;
        Seg.set(i, v);
    }
    Seg.build();

    int Q;
    cin >> Q;
    while (Q--) {
        int l, r;
        cin >> l >> r;
        cout << Seg.query(l - 1, r - 1) << '\n';
    }
}